### PR TITLE
Add role management API

### PR DESCRIPTION
**Pull request by ashensw**

### CODE REVIEW

File: modules/charon-core/src/main/java/org/wso2/charon3/core/objects/Role.java [hunk 1]
```diff
@@ -0,0 +1,352 @@
...
+     *
+     * @return Display name of the role.
+     */
+    public String getDisplayName() {
+
+        if (isAttributeExist(SCIMConstants.RoleSchemaConstants.DISPLAY_NAME)) {
+            return LambdaExceptionUtils.rethrowSupplier(
+                    () -> ((SimpleAttribute) attributeList.get(SCIMConstants.RoleSchemaConstants.DISPLAY_NAME))
+                            .getStringValue()).get();
+        } else {
```
Comment (dewniMW) [inline modules/charon-core/src/main/java/org/wso2/charon3/core/objects/Role.java:56]: no need to have `else`

File: modules/charon-core/src/main/java/org/wso2/charon3/core/objects/Role.java [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+    private ComplexAttribute processUser(User user) throws BadRequestException, CharonException {
+
+        String userId = user.getId();
+        String userName = user.getUserName();
+        String reference = user.getLocation();
+
+        ComplexAttribute complexAttribute = new ComplexAttribute();
+        complexAttribute.setName(SCIMConstants.RoleSchemaConstants.USERS + "_" + userId + SCIMConstants.DEFAULT);
+
+        if (userId != null) {
```
Comment (dewniMW) [inline modules/charon-core/src/main/java/org/wso2/charon3/core/objects/Role.java:188]: shall we use `StringUtils.isNotBlank` ?

File: modules/charon-core/src/main/java/org/wso2/charon3/core/objects/Role.java [hunk 3]
```diff
@@ -0,0 +1,352 @@
...
+                List<Attribute> subValuesList = groups.getAttributeValues();
+                for (Attribute subValue : subValuesList) {
+                    ComplexAttribute complexAttribute = (ComplexAttribute) subValue;
+                    Map<String, Attribute> subAttributesList = complexAttribute.getSubAttributesList();
+                    if (subAttributesList != null && subAttributesList
+                            .containsKey(SCIMConstants.CommonSchemaConstants.DISPLAY)) {
+                        displayNames.add((String) ((SimpleAttribute) (subAttributesList
+                                .get(SCIMConstants.CommonSchemaConstants.DISPLAY))).getValue());
+                    }
+
```
Comment (dewniMW) [inline modules/charon-core/src/main/java/org/wso2/charon3/core/objects/Role.java:263]: remove the extra line here

File: modules/charon-core/src/main/java/org/wso2/charon3/core/extensions/RoleManager.java [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+     *
+     * @param id                 Role ID.
+     * @param requiredAttributes Required Attributes.
+     * @return Role.
+     * @throws NotImplementedException NotImplementedException.
+     * @throws BadRequestException     BadRequestException.
+     * @throws CharonException         CharonException.
+     * @throws NotFoundException       NotFoundException.
+     */
+    Role getRole(String id, Map<String, Boolean> requiredAttributes)
```
Comment (emswbandara) [inline modules/charon-core/src/main/java/org/wso2/charon3/core/extensions/RoleManager.java:61]: shouldn't we support excludeAttributes? it is defined for users/groups in SCIM spec

File: modules/charon-core/src/main/java/org/wso2/charon3/core/extensions/RoleManager.java [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+     * @param startIndex Start Index
+     * @param count      Count
+     * @param sortBy     Sort by
+     * @param sortOrder  Sort order
+     * @return List of roles.
+     * @throws CharonException         CharonException.
+     * @throws NotImplementedException NotImplementedException.
+     * @throws BadRequestException     BadRequestException.
+     */
+    List<Object> listRolesWithGET(Node node, Integer startIndex, Integer count, String sortBy, String sortOrder)
```
Comment (emswbandara) [inline modules/charon-core/src/main/java/org/wso2/charon3/core/extensions/RoleManager.java:88]: Normally for other APIs we supported requiredAttributes during list operation. shouldn't that be the case?

File: modules/charon-core/src/main/java/org/wso2/charon3/core/extensions/RoleManager.java [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+     *
+     * @param oldRole Old role.
+     * @param newRole new role.
+     * @return Updated role.
+     * @throws NotImplementedException NotImplementedException.
+     * @throws BadRequestException     BadRequestException.
+     * @throws CharonException         CharonException.
+     * @throws NotFoundException       NotFoundException.
+     */
+    Role updateRole(Role oldRole, Role newRole)
```
Comment (emswbandara) [inline modules/charon-core/src/main/java/org/wso2/charon3/core/extensions/RoleManager.java:102]: shouldn't we support updateRole using role ID?

File: modules/charon-core/src/main/java/org/wso2/charon3/core/extensions/RoleManager.java [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+    /**
+     * List roles with Post.
+     *
+     * @param searchRequest Search request.
+     * @return List of roles.
+     * @throws NotImplementedException NotImplementedException.
+     * @throws BadRequestException     BadRequestException.
+     * @throws CharonException         CharonException.
+     */
+    List<Object> listRolesWithPost(SearchRequest searchRequest)
```
Comment (emswbandara) [inline modules/charon-core/src/main/java/org/wso2/charon3/core/extensions/RoleManager.java:114]: shouldn't we support pagination here?

File: modules/charon-core/src/main/java/org/wso2/charon3/core/objects/Role.java (hunk 3 above)
Comment (ashensw) [reply modules/charon-core/src/main/java/org/wso2/charon3/core/objects/Role.java]: Fixed and updated the PR.

File: modules/charon-core/src/main/java/org/wso2/charon3/core/objects/Role.java (hunk 2 above)
Comment (ashensw) [reply modules/charon-core/src/main/java/org/wso2/charon3/core/objects/Role.java:187]: StringUtils is not used by the charon repo. Therefore thought of not adding the dependency as it's not really necessary for this particular case.

File: modules/charon-core/src/main/java/org/wso2/charon3/core/extensions/RoleManager.java (hunk 5 above)
Comment (ashensw) [reply modules/charon-core/src/main/java/org/wso2/charon3/core/extensions/RoleManager.java:88]: `listRoles` APIs will only have basic roles information without additional properties such as users, groups etc. Therefore we decided not to include `attributes` and `excludeAttributes` here and finalized the swagger definition accordingly.

File: modules/charon-core/src/main/java/org/wso2/charon3/core/extensions/RoleManager.java (hunk 6 above)
Comment (ashensw) [reply modules/charon-core/src/main/java/org/wso2/charon3/core/extensions/RoleManager.java:102]: ID is included inside the `Role` object.

File: modules/charon-core/src/main/java/org/wso2/charon3/core/extensions/RoleManager.java (hunk 7 above)
Comment (ashensw) [reply modules/charon-core/src/main/java/org/wso2/charon3/core/extensions/RoleManager.java:114]: Pagination parameters are included inside the `searchRequest`.

File: modules/charon-core/src/main/java/org/wso2/charon3/core/protocol/endpoints/RoleResourceManager.java [hunk 8]
```diff
@@ -0,0 +1,543 @@
...
+        try {
+            if (roleManager == null) {
+                String error = "Provided role manager handler is null.";
+                throw new InternalErrorException(error);
+            }
+            JSONEncoder encoder = getEncoder();
+            SCIMResourceTypeSchema schema = SCIMResourceSchemaManager.getInstance().getRoleResourceSchema();
+            Map<String, Boolean> requestAttributes = ResourceManagerUtil.getAllAttributeURIs(schema);
+
+            Role oldRole = roleManager.getRole(id, requestAttributes);
```
Comment (emswbandara) [inline modules/charon-core/src/main/java/org/wso2/charon3/core/protocol/endpoints/RoleResourceManager.java:386]: will this retrieve the entire user list and group list of role?

Comment (ashensw) [reply modules/charon-core/src/main/java/org/wso2/charon3/core/protocol/endpoints/RoleResourceManager.java:386]: Yes. since it needed to compare.

Comment (emswbandara) [reply modules/charon-core/src/main/java/org/wso2/charon3/core/protocol/endpoints/RoleResourceManager.java:386]: let's create a git issue to track the perf improvement

**Pull request merged**